From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Eldar Rello <eldar.rello@gmail.com>
Date: Tue, 13 Apr 2021 23:26:23 +0300
Subject: Make Opus complexity configurable with command line switch

Make Opus complexity configurable with command line switch

Signed-off-by: gellert <gellert.hegyi@around.co>

diff --git a/audio/channel_send.cc b/audio/channel_send.cc
index bbfffef1a1fd2c9f3e5d894925a83bd67ff04c30..bae1dee0077af23ca60342f4043353df475513a8 100644
--- a/audio/channel_send.cc
+++ b/audio/channel_send.cc
@@ -536,6 +536,7 @@ ChannelSend::ChannelSend(
       HasSwitch(switches::kDisableEncoderQueue)) {
     use_encoder_queue_ = false;
   }
+  (void)worker_thread_checker_;
 }
 
 ChannelSend::~ChannelSend() {
diff --git a/modules/audio_coding/BUILD.gn b/modules/audio_coding/BUILD.gn
index 3ca2e9915389bed7669b6335b6b63cff43244899..cee23a5d09c2dd6f4318bea26d2d2c8c63fbdeaf 100644
--- a/modules/audio_coding/BUILD.gn
+++ b/modules/audio_coding/BUILD.gn
@@ -745,6 +745,7 @@ rtc_library("webrtc_opus") {
   ]
 
   deps = [
+    "//media:media_buildflags",
     ":audio_coding_opus_common",
     ":audio_network_adaptor",
     "../../api:array_view",
diff --git a/modules/audio_coding/codecs/opus/audio_encoder_opus.cc b/modules/audio_coding/codecs/opus/audio_encoder_opus.cc
index 5b3b7dab19d6d44e79276ddc3c934ad44414c34a..c72e0d92cbbdcc54a591db4dfd0dc620e6323db2 100644
--- a/modules/audio_coding/codecs/opus/audio_encoder_opus.cc
+++ b/modules/audio_coding/codecs/opus/audio_encoder_opus.cc
@@ -17,6 +17,9 @@
 #include <utility>
 
 #include "absl/strings/match.h"
+#include "base/command_line.h"
+#include "base/strings/string_number_conversions.h"
+#include "media/base/media_switches.h"
 #include "modules/audio_coding/audio_network_adaptor/audio_network_adaptor_impl.h"
 #include "modules/audio_coding/audio_network_adaptor/controller_manager.h"
 #include "modules/audio_coding/codecs/opus/audio_coder_opus_common.h"
@@ -253,7 +256,11 @@ absl::optional<AudioEncoderOpusConfig> AudioEncoderOpusImpl::SdpToConfig(
   config.application = config.num_channels == 1
                            ? AudioEncoderOpusConfig::ApplicationMode::kVoip
                            : AudioEncoderOpusConfig::ApplicationMode::kAudio;
-
+  int opus_complexity;
+  if (base::StringToInt(base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII(
+      switches::kOpusComplexity), &opus_complexity)) {
+    config.complexity = opus_complexity;
+  }
   constexpr int kMinANAFrameLength = kANASupportedFrameLengths[0];
   constexpr int kMaxANAFrameLength =
       kANASupportedFrameLengths[arraysize(kANASupportedFrameLengths) - 1];
