From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Heilig Benedek <benecene@gmail.com>
Date: Tue, 8 Oct 2019 21:55:21 +0200
Subject: audioworklet_realtime_thread_default_true


diff --git a/third_party/blink/renderer/platform/scheduler/common/thread.cc b/third_party/blink/renderer/platform/scheduler/common/thread.cc
index b657a3c80c1c47547c69c4d55c553305e52c3e0f..c11fb1e90c3c0f7980ca3b1f087875468de8ab94 100644
--- a/third_party/blink/renderer/platform/scheduler/common/thread.cc
+++ b/third_party/blink/renderer/platform/scheduler/common/thread.cc
@@ -90,10 +90,13 @@ std::unique_ptr<Thread> Thread::CreateWebAudioThread() {
   // WebAudio uses a thread with |DISPLAY| priority to avoid glitch when the
   // system is under the high pressure. Note that the main browser thread also
   // runs with same priority. (see: crbug.com/734539)
-  params.thread_priority =
-      base::FeatureList::IsEnabled(features::kAudioWorkletRealtimeThread)
-          ? base::ThreadPriority::REALTIME_AUDIO
-          : base::ThreadPriority::DISPLAY;
+  // params.thread_priority =
+  //     base::FeatureList::IsEnabled(features::kAudioWorkletRealtimeThread)
+  //         ? base::ThreadPriority::REALTIME_AUDIO
+  //         : base::ThreadPriority::DISPLAY;
+
+  // @AROUND: we always use real-time priority for the AudioWorklet thread
+  params.thread_priority = base::ThreadPriority::REALTIME_AUDIO;

   return CreateThread(params);
 }
