From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Sat, 16 May 2020 00:11:23 +0200
Subject: disables goog agc

Disables experimental AGC1 mode in MediaStreamAudioProcessor.

Signed-off-by: gellert <gellert.hegyi@around.co>

diff --git a/third_party/blink/renderer/platform/mediastream/media_stream_audio_processor_options.cc b/third_party/blink/renderer/platform/mediastream/media_stream_audio_processor_options.cc
index 2cbfffdd31d373740dd81aabd31fa4dd955d96e5..d35b9f2305adb2e828fee2dedb5fa752d6a094ba 100644
--- a/third_party/blink/renderer/platform/mediastream/media_stream_audio_processor_options.cc
+++ b/third_party/blink/renderer/platform/mediastream/media_stream_audio_processor_options.cc
@@ -438,7 +438,9 @@ rtc::scoped_refptr<webrtc::AudioProcessing> CreateWebRtcAudioProcessingModule(
     experimental_agc->digital_adaptive_disabled =
         hybrid_agc_params.has_value() && !hybrid_agc_params->dry_run;
     config.Set<webrtc::ExperimentalAgc>(experimental_agc);
-#if BUILDFLAG(IS_CHROMECAST)
+// @AROUND we want the Experimental AGC option to always be turned off if
+// it's disabled in the APM properties (media_stream_audio_processor_options.cc)
+// #if defined(IS_CHROMECAST)
   } else {
     // Do not use the analog controller.
     // This should likely be done on non-Chromecast platforms as well, but care
@@ -446,7 +448,7 @@ rtc::scoped_refptr<webrtc::AudioProcessing> CreateWebRtcAudioProcessingModule(
     // https://crbug.com/918677#c4
     config.Set<webrtc::ExperimentalAgc>(
         new webrtc::ExperimentalAgc(/*enabled=*/false));
-#endif  // BUILDFLAG(IS_CHROMECAST)
+// #endif  // BUILDFLAG(IS_CHROMECAST)
   }
 
   // Create and configure the webrtc::AudioProcessing.
