From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Heilig Benedek <benecene@gmail.com>
Date: Mon, 21 Oct 2019 11:31:54 +0200
Subject: disable_bgcolor_dcheck


diff --git a/content/browser/renderer_host/render_widget_host_view_base.cc b/content/browser/renderer_host/render_widget_host_view_base.cc
index 3dc776ac982fdd5d4d15bd3afdf208810340db4e..0ce2017496f94a321c850a936fbe769fe9c0f4b4 100644
--- a/content/browser/renderer_host/render_widget_host_view_base.cc
+++ b/content/browser/renderer_host/render_widget_host_view_base.cc
@@ -303,8 +303,8 @@ void RenderWidgetHostViewBase::SetBackgroundColor(SkColor color) {
   // TODO(danakj): OPAQUE colors only make sense for main frame widgets,
   // as child frames are always transparent background. We should move this to
   // RenderView instead.
-  DCHECK(SkColorGetA(color) == SK_AlphaOPAQUE ||
-         SkColorGetA(color) == SK_AlphaTRANSPARENT);
+  // DCHECK(SkColorGetA(color) == SK_AlphaOPAQUE ||
+  //        SkColorGetA(color) == SK_AlphaTRANSPARENT);
   if (default_background_color_ == color)
     return;
 
